examples/step-72/step-72.cc

```diff
@@ -0,0 +1,1147 @@
+
+    sparsity_pattern.copy_from(dsp);
+    system_matrix.reinit(sparsity_pattern);
+  }
+
+  // @sect4{Assembling the linear system}
+
+  // @sect5{Manual assembly}
+
+  // The assembly functions are the interesting contributions to this tutorial.
+  // The assemble_system_unassisted() method implements exactly the same
+  // assembly function as is detailed in step-15, but in this instance we
+  // use the MeshWorker::mesh_loop() function to multithread the assembly
+  // process. The reason for doing this is quite simple: When using
+  // automatic differentiation, we know that there is to be some additional
+  // computational overhead incurred. In order to mitigate this performance
+  // loss, we'd like to take advantage of as many (easily available)
+  // computational resources as possible. The MeshWorker::mesh_loop() concept
+  // makes this a relatively straightforward task. At the same time, for the
+  // purposes of fair comparison, we need to do the same to the implementation
+  // that uses no assistance when computing the residual or its linearization.
+  // (The MeshWorker::mesh_loop() function is first discussed in step-12 and
+  // step-16, if you'd like to read up on it.)
+  //
+  // The steps required to implement the multithreading are the same between the
+  // three functions, so we'll use the assemble_system_unassisted() function
+  // as an opportunity to focus on the multithreading itself.
+  template <int dim>
+  void MinimalSurfaceProblem<dim>::assemble_system_unassisted()
+  {
+    system_matrix = 0;
+    system_rhs    = 0;
+
+    const unsigned int dofs_per_cell = fe.n_dofs_per_cell();
+
+    // The MeshWorker::mesh_loop() expects that we provide two exemplar data
+    // structures. The first, `ScratchData`, is to store all large data that
+    // is to be reused between threads. The `CopyData` will hold the
+    // contributions to the linear system that come from each cell. These
+    // independent matrix-vector pairs must be accumulated into the
+    // global linear system sequentially. Since we don't need anything
+    // on top of what the MeshWorker::ScratchData and MeshWorker::CopyData
+    // classes already provide, we use these exact class definitions for
+    // our problem. Note that we only require a single instance of a local
+    // matrix, local right-hand side vector, and cell degree of freedom index
+    // vector -- the MeshWorker::CopyData therefore has `1` for all three
+    // of its template arguments.
+    using ScratchData = MeshWorker::ScratchData<dim>;
+    using CopyData    = MeshWorker::CopyData<1, 1, 1>;
+
+    // We also need to know what type of iterator we'll be working with
+    // during assembly. For simplicity, we just ask the compiler to work
+    // this out for us using the decltype() specifier, knowing that we'll
+    // be iterating over active cells owned by the @p dof_handler.
+    using CellIteratorType = decltype(dof_handler.begin_active());
```

Comment:
Why not write below simply:
```cpp
    auto cell_worker = [this](const auto &cell,
                              auto &      scratch_data,
                              auto &      copy_data) {
```
?

Reply:
I'm not a huge fan of `auto` for lambda functions unless the type is clear to a reader -- which I don't think it is. I'd much rather be explicit:
```suggestion
    using CellIteratorType = typename DoFHandler<dim>::active_cell_iterator;
```

Reply:
But then I'm not the author of the program and am happy to defer.

Reply:
The way that I wrote it follows exactly the pattern that we outline in the documentation for `mesh_loop`. See [here](https://dealii.org/developer/doxygen/deal.II/group__MeshWorker.html#ga76ec61fbd188fb320fe8ca166a79b322). I'd therefore prefer to keep it as it is, if you all don't mind.